crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1654,48 +1654,56 @@ impl<'db> TypeInferenceBuilder<'db> {
     fn infer_fstring_expression(&mut self, fstring: &ast::ExprFString) -> Type<'db> {
         let ast::ExprFString { range: _, value } = fstring;
 
+        let mut has_expression = false;
+        let mut literals = Vec::new();
+        // Build an arena to allocate expression string representation
+        let expr_arena = typed_arena::Arena::new();
```

Comment:
Could we, instead of using a `literals` `Vec` and an `expr_arena` instead use a `String` and concatenated the parts in-place? It also allows you to bail out early as soon as the string exceeds the max string literal size instead of iterating to the end.

Reply:
I think my intention was to concatenate last minute so very long `f-string` with a variable interpolation fairly late are not allocated (we keep everything as `Box<str>` as long as possible).
There is the problem of allocating `Int` / `Bool` repr, but they realistically should never be too big (unless a pathological case is crafted on purpose).

Reply:
Implemented in 3ee3fcbc3bacc40df0bf13dbdefdcc57c7655757 if you want to compare versions

Reply:
> Implemented in [3ee3fcb](https://github.com/astral-sh/ruff/commit/3ee3fcbc3bacc40df0bf13dbdefdcc57c7655757) if you want to compare versions

Thanks. I do like that better